src/VisualStudio/CSharp/Impl/LanguageService/CSharpCodeCleanupFixer.cs

```diff
@@ -0,0 +1,157 @@
+                                                                           IDEDiagnosticIds.RemoveUnnecessaryCastDiagnosticId,
+                                                                           IDEDiagnosticIds.UseExpressionBodyForConstructorsDiagnosticId,
+                                                                           IDEDiagnosticIds.UseExpressionBodyForMethodsDiagnosticId,
+                                                                           IDEDiagnosticIds.UseExpressionBodyForConversionOperatorsDiagnosticId,
+                                                                           IDEDiagnosticIds.UseExpressionBodyForOperatorsDiagnosticId,
+                                                                           IDEDiagnosticIds.UseExpressionBodyForPropertiesDiagnosticId,
+                                                                           IDEDiagnosticIds.UseExpressionBodyForIndexersDiagnosticId,
+                                                                           IDEDiagnosticIds.UseExpressionBodyForAccessorsDiagnosticId,
+                                                                           IDEDiagnosticIds.InlineDeclarationDiagnosticId,
+                                                                           CSharpRemoveUnusedVariableCodeFixProvider.CS0168,
+                                                                           CSharpRemoveUnusedVariableCodeFixProvider.CS0219,
+                                                                           IDEDiagnosticIds.UseObjectInitializerDiagnosticId,
+                                                                           IDEDiagnosticIds.UseCollectionInitializerDiagnosticId);
+        
+        public CSharpCodeCleanupFixer(ICodeFixService codeFixService, IThreadingContext threadingContext, bool assertIsForeground = false)
+            : base(threadingContext, assertIsForeground)
+        {
+            _codeFixServiceOpt = codeFixService;
+        }
+
+        public async Task<bool> FixAsync(ICodeCleanUpScope scope, FixIdContainer enabledFixIds, CancellationToken cancellationToken)
+        {
+            var textBufferScope = scope as TextBufferCodeCleanUpScope;
+            if (textBufferScope == null)
+            {
+                var hierarchyContent = scope as IVsHierarchyCodeCleanupScope;
+                var hierarchy = hierarchyContent.Hierarchy;
+                var itemId = hierarchyContent.ItemId;
+
+                if (hierarchy.GetCanonicalName(itemId, out var path) == 0)
+                {
+                    var attr = File.GetAttributes(path);
+                    if (attr.HasFlag(FileAttributes.Directory))
+                    {
+                        // directory
+                    }
+                    else
+                    {   
+                        // document
+                    }
+                }
```

Comment:
wow... this is really surprising.  Is this a new API?  Why is it built on top of tehse primitives/concepts?

this also basically forces thsi system to *have* to be at the VS layer, instead of being at the feature layer where it could be used by a host like VSCode.

Reply:
Presumably this editor API will be defined in an assembly that can be moved to a new environment. /cc @olegtk